// @armoriq/sdk/dist/config.js
"use strict";
/**
 * armoriq.yaml configuration models and loader.
 *
 * Mirrors armoriq_sdk/config.py shape so the same file is portable
 * between the two SDKs.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArmorIQConfigError = exports.DEFAULT_PROXY_URL = void 0;
exports.resolveEnvReference = resolveEnvReference;
exports.parseArmorIQConfig = parseArmorIQConfig;
exports.loadArmorIQConfig = loadArmorIQConfig;
exports.resolveEnvReferences = resolveEnvReferences;
exports.saveArmorIQConfig = saveArmorIQConfig;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const yaml = __importStar(require("js-yaml"));
exports.DEFAULT_PROXY_URL = 'https://customer-proxy.armoriq.ai';
class ArmorIQConfigError extends Error {
    constructor(message) {
        super(message);
        this.name = 'ArmorIQConfigError';
    }
}
exports.ArmorIQConfigError = ArmorIQConfigError;
/**
 * Resolve `$ENV_VAR` or `${ENV_VAR}` references against process.env.
 * Returns the original string when it is not an env reference.
 */
function resolveEnvReference(value) {
    if (typeof value !== 'string')
        return value;
    if (value.startsWith('${') && value.endsWith('}')) {
        return process.env[value.slice(2, -1)] ?? '';
    }
    if (value.startsWith('$') && value.length > 1 && !value.includes(' ')) {
        return process.env[value.slice(1)] ?? '';
    }
    return value;
}
function normalizeAuth(value) {
    if (value === null || value === undefined)
        return { type: 'none' };
    if (typeof value === 'string')
        return { type: value };
    return value;
}
function validateIdentity(raw) {
    if (!raw || typeof raw !== 'object') {
        throw new ArmorIQConfigError('identity is required');
    }
    for (const k of ['api_key', 'user_id', 'agent_id']) {
        if (typeof raw[k] !== 'string' || !raw[k]) {
            throw new ArmorIQConfigError(`identity.${k} is required`);
        }
    }
    return { api_key: raw.api_key, user_id: raw.user_id, agent_id: raw.agent_id };
}
function validateAuth(raw) {
    const a = normalizeAuth(raw);
    if (a.type === 'bearer' && !a.token) {
        throw new ArmorIQConfigError("bearer auth requires 'token'");
    }
    if (a.type === 'api_key' && !a.api_key) {
        throw new ArmorIQConfigError("api_key auth requires 'api_key'");
    }
    if (a.type !== 'bearer' && a.type !== 'api_key' && a.type !== 'none') {
        throw new ArmorIQConfigError(`unknown auth type: ${a.type}`);
    }
    return a;
}
function parseArmorIQConfig(raw) {
    if (!raw || typeof raw !== 'object') {
        throw new ArmorIQConfigError('config root must be an object');
    }
    const version = raw.version ?? 'v1';
    if (version !== 'v1') {
        throw new ArmorIQConfigError(`unsupported config version: ${version}`);
    }
    const environment = raw.environment ?? 'sandbox';
    if (environment !== 'sandbox' && environment !== 'production') {
        throw new ArmorIQConfigError(`environment must be sandbox or production, got: ${environment}`);
    }
    const proxy = {
        url: raw.proxy?.url ?? exports.DEFAULT_PROXY_URL,
        timeout: raw.proxy?.timeout ?? 30,
        max_retries: raw.proxy?.max_retries ?? 3,
    };
    const policy = {
        allow: Array.isArray(raw.policy?.allow) ? raw.policy.allow : [],
        deny: Array.isArray(raw.policy?.deny) ? raw.policy.deny : [],
    };
    const intent = {
        ttl_seconds: raw.intent?.ttl_seconds ?? 300,
        require_csrg: raw.intent?.require_csrg ?? true,
    };
    const mcpServersRaw = Array.isArray(raw.mcp_servers) ? raw.mcp_servers : [];
    const mcpServers = mcpServersRaw.map((s, i) => {
        if (!s.id || typeof s.id !== 'string') {
            throw new ArmorIQConfigError(`mcp_servers[${i}].id is required`);
        }
        if (!s.url || typeof s.url !== 'string') {
            throw new ArmorIQConfigError(`mcp_servers[${i}].url is required`);
        }
        return {
            id: s.id,
            url: s.url,
            description: s.description,
            auth: validateAuth(s.auth),
        };
    });
    const ids = mcpServers.map((s) => s.id);
    if (new Set(ids).size !== ids.length) {
        throw new ArmorIQConfigError("mcp_servers contain duplicate 'id' values");
    }
    return {
        version: 'v1',
        identity: validateIdentity(raw.identity),
        environment,
        proxy,
        mcp_servers: mcpServers,
        policy,
        intent,
    };
}
function loadArmorIQConfig(filePath = 'armoriq.yaml') {
    const resolved = path.resolve(filePath);
    if (!fs.existsSync(resolved)) {
        throw new ArmorIQConfigError(`Config file not found: ${resolved}`);
    }
    const raw = yaml.load(fs.readFileSync(resolved, 'utf-8')) ?? {};
    return parseArmorIQConfig(raw);
}
/**
 * Resolve every `$VAR` reference in the config against process.env.
 * Returns a new config with the references replaced by their values.
 */
function resolveEnvReferences(config) {
    return {
        ...config,
        identity: {
            api_key: resolveEnvReference(config.identity.api_key),
            user_id: resolveEnvReference(config.identity.user_id),
            agent_id: resolveEnvReference(config.identity.agent_id),
        },
        mcp_servers: config.mcp_servers.map((s) => ({
            ...s,
            url: resolveEnvReference(s.url),
            auth: s.auth.type === 'bearer'
                ? { type: 'bearer', token: resolveEnvReference(s.auth.token ?? '') }
                : s.auth.type === 'api_key'
                    ? { type: 'api_key', api_key: resolveEnvReference(s.auth.api_key ?? '') }
                    : s.auth,
        })),
    };
}
function saveArmorIQConfig(config, filePath = 'armoriq.yaml') {
    const data = {
        version: config.version,
        identity: { ...config.identity },
        environment: config.environment,
        proxy: { ...config.proxy },
        mcp_servers: config.mcp_servers.map((s) => {
            const out = {
                id: s.id,
                url: s.url,
                auth: s.auth.type === 'none'
                    ? 'none'
                    : s.auth.type === 'bearer'
                        ? { type: 'bearer', token: s.auth.token }
                        : { type: 'api_key', api_key: s.auth.api_key },
            };
            if (s.description)
                out.description = s.description;
            return out;
        }),
        policy: { allow: config.policy.allow, deny: config.policy.deny },
        intent: { ttl_seconds: config.intent.ttl_seconds, require_csrg: config.intent.require_csrg },
    };
    fs.writeFileSync(filePath, yaml.dump(data, { sortKeys: false }), 'utf-8');
}
