// @armoriq/sdk/dist/plan_builder.js
"use strict";
/**
 * Plan-shape helpers used by ArmorIQSession and framework integrations.
 *
 * Most LLM frameworks surface tool calls as a flat list of {name, args}.
 * The SDK accepts that shape directly so plugin code doesn't have to
 * hand-construct {goal, steps: [...]} every time.
 *
 * Mirrors armoriq_sdk/plan_builder.py — outputs the same plan shape and
 * hash digest so PY and TS sides round-trip cleanly.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultToolNameParser = defaultToolNameParser;
exports.buildPlanFromToolCalls = buildPlanFromToolCalls;
exports.hashToolCalls = hashToolCalls;
const crypto = __importStar(require("crypto"));
/**
 * Default tool-name convention: `<MCP>__<action>`. Matches the proxy's
 * MCP gateway and the convention used by sdk-admin-agent.
 *
 *   "Stripe__create_payment" -> { mcp: "Stripe", action: "create_payment" }
 *   "create_payment"          -> uses defaultMcpName, else throws.
 */
function defaultToolNameParser(defaultMcpName) {
    return (toolName) => {
        const idx = toolName.indexOf('__');
        if (idx === -1) {
            if (!defaultMcpName) {
                throw new Error(`Tool "${toolName}" is not namespaced as <MCP>__<action> and no defaultMcpName was set on the session.`);
            }
            return { mcp: defaultMcpName, action: toolName };
        }
        const mcp = toolName.slice(0, idx);
        const action = toolName.slice(idx + 2);
        if (!mcp || !action) {
            throw new Error(`Tool "${toolName}" has a malformed MCP prefix.`);
        }
        return { mcp, action };
    };
}
function asToolCall(tc) {
    return { name: tc.name, args: tc.args ?? {} };
}
/**
 * Build an SDK-shaped plan dict from a flat list of tool calls.
 */
function buildPlanFromToolCalls(toolCalls, goal, toolNameParser, defaultMcpName) {
    const parser = toolNameParser ?? defaultToolNameParser(defaultMcpName);
    const steps = toolCalls.map((tc) => {
        const call = asToolCall(tc);
        const { mcp, action } = parser(call.name);
        return {
            action,
            tool: action,
            mcp,
            params: call.args ?? {},
            description: `Call ${action} on ${mcp}`,
        };
    });
    return { goal: goal ?? 'agent task', steps };
}
/**
 * Stable hash over a tool-calls list — used by ArmorIQSession to skip
 * re-minting when the LLM re-emits the same plan in the same turn.
 *
 * Uses JSON.stringify with no key sorting, matching the Python side's
 * json.dumps(separators=(",",":")), so TS and PY produce matching digests.
 */
function hashToolCalls(toolCalls) {
    const canonicalList = toolCalls.map((tc) => ({
        name: asToolCall(tc).name,
        args: asToolCall(tc).args ?? {},
    }));
    const canonical = JSON.stringify(canonicalList);
    return crypto.createHash('sha256').update(canonical, 'utf8').digest('hex');
}
