// @armoriq/sdk/dist/integrations/google_adk.js
"use strict";
/**
 * ArmorIQ — Google ADK integration (TypeScript).
 *
 * Mirrors armoriq_sdk/integrations/google_adk.py.
 *
 * Multi-user pattern:
 *
 *     import { ArmorIQADK } from '@armoriq/sdk/integrations/google_adk';
 *     const armoriq = new ArmorIQADK({ apiKey: process.env.ARMORIQ_API_KEY! });
 *
 *     // per-request, inside your chat handler:
 *     const scope = await armoriq.forUser(userEmail, { goal: message });
 *     scope.install(rootAgent);
 *     try {
 *       for await (const event of runner.runAsync(...)) { ... }
 *     } finally {
 *       scope.uninstall(rootAgent);
 *     }
 *
 * The bundle installs three ADK lifecycle callbacks on the agent:
 *   afterModelCallback  → mint intent token from LLM tool calls
 *   beforeToolCallback  → enforce per-user policy, block/hold if needed
 *   afterToolCallback   → audit report
 *
 * "Agent" here is duck-typed — anything with three settable callback
 * properties works. We don't take a hard dep on a specific TS ADK
 * package because there isn't a single dominant one yet; the contract
 * is a bag of three async functions.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArmorIQADKBundle = exports.ArmorIQADK = void 0;
const client_1 = require("../client");
/**
 * Process-wide ArmorIQ factory for ADK-style agents.
 */
class ArmorIQADK {
    client;
    mode;
    llm;
    validitySeconds;
    defaultMcpName;
    // autoReanchor flag removed 2026-05-13 — always on. Property retained
    // as `readonly = true` so any caller reading factory.autoReanchor still
    // observes the same value, but no env or option toggles it.
    autoReanchor = true;
    trueReanchor;
    customParser;
    bootstrapData;
    explicitAgentName;
    constructor(opts) {
        this.explicitAgentName = opts.agentName;
        this.client = new client_1.ArmorIQClient({
            apiKey: opts.apiKey,
            backendEndpoint: opts.backendEndpoint,
            iapEndpoint: opts.iapEndpoint ?? opts.backendEndpoint,
            proxyEndpoint: opts.proxyEndpoint ?? opts.backendEndpoint,
            useProduction: opts.useProduction ?? true,
            userId: 'agent',
            // When agentName is not explicitly provided, omit agentId so the
            // client constructor's env fallback (AGENT_ID) can kick in.
            ...(opts.agentName ? { agentId: opts.agentName } : {}),
        });
        this.defaultMcpName = opts.defaultMcpName;
        this.customParser = opts.toolNameParser;
        this.validitySeconds = opts.validitySeconds ?? 300;
        this.mode = opts.mode ?? 'sdk';
        this.llm = opts.llm ?? 'agent';
        this.trueReanchor = opts.trueReanchor ?? true;
    }
    async bootstrap() {
        if (!this.bootstrapData) {
            this.bootstrapData = await this.client.bootstrap();
            const orgName = this.bootstrapData.org?.name ?? 'unknown';
            // Auto-resolve agentId from bootstrap only when the caller didn't
            // pass an explicit agentName. The agents[] list (new in bootstrap
            // response) is authoritative — the legacy agent.name returned the
            // API key's display name which caused cross-agent policy
            // attribution issues (#199).
            let resolvedAgentName;
            if (!this.explicitAgentName) {
                const agents = this.bootstrapData.agents ?? [];
                const single = agents.length === 1 ? agents[0] : null;
                const fallback = this.bootstrapData.agent?.name;
                resolvedAgentName = single?.name ?? fallback;
                if (resolvedAgentName) {
                    this.client._setAgentId(resolvedAgentName);
                }
            }
            const mcps = Array.isArray(this.bootstrapData.mcps)
                ? this.bootstrapData.mcps.map((m) => m.name)
                : [];
            const toolMapSize = Object.keys(this.bootstrapData.toolMap ?? {}).length;
            const loggedAgentName = this.explicitAgentName ?? resolvedAgentName ?? 'unknown';
            console.info(`[armoriq] bootstrap: org=${orgName} agent=${loggedAgentName} mcps=${JSON.stringify(mcps)} toolMap=${toolMapSize}`);
        }
        return this.bootstrapData;
    }
    async toolNameParser() {
        if (this.customParser)
            return this.customParser;
        const bootstrap = await this.bootstrap();
        const toolMap = bootstrap.toolMap ?? {};
        const defaultMcp = this.defaultMcpName;
        return (toolName) => {
            const mcp = toolMap[toolName];
            if (mcp)
                return { mcp, action: toolName };
            if (toolName.includes('__')) {
                const idx = toolName.indexOf('__');
                return { mcp: toolName.slice(0, idx), action: toolName.slice(idx + 2) };
            }
            if (defaultMcp)
                return { mcp: defaultMcp, action: toolName };
            return { mcp: 'unknown', action: toolName };
        };
    }
    invalidateUser(userEmail) {
        this.client.invalidateUser(userEmail);
    }
    async forUser(userEmail, opts) {
        await this.bootstrap();
        const scope = this.client.forUser(userEmail);
        const parser = await this.toolNameParser();
        return new ArmorIQADKBundle({
            factory: this,
            scope,
            parser,
            userEmail: userEmail.trim().toLowerCase(),
            goal: opts?.goal ?? 'agent task',
        });
    }
}
exports.ArmorIQADK = ArmorIQADK;
/**
 * Per-request ADK bundle — installs/uninstalls lifecycle callbacks on
 * one agent and binds them to one user's session.
 */
class ArmorIQADKBundle {
    userEmail;
    goal;
    session;
    factory;
    scope;
    parser;
    planMinted = false;
    blockedTools = new Set();
    blockedActions = new Map();
    agent;
    saved = {};
    constructor(args) {
        this.factory = args.factory;
        this.scope = args.scope;
        this.parser = args.parser;
        this.userEmail = args.userEmail;
        this.goal = args.goal;
    }
    ensureSession() {
        if (!this.session) {
            const opts = {
                mode: this.factory.mode,
                validitySeconds: this.factory.validitySeconds,
                llm: this.factory.llm,
                toolNameParser: this.parser,
                defaultMcpName: this.factory.defaultMcpName,
                trueReanchor: this.factory.trueReanchor,
            };
            this.session = this.scope.startSession(opts);
        }
        return this.session;
    }
    async afterModel(_callbackContext, llmResponse) {
        try {
            if (this.planMinted)
                return null;
            const parts = llmResponse?.content?.parts ?? [];
            const toolCalls = [];
            for (const p of parts) {
                const fc = p?.functionCall ?? p?.function_call;
                if (fc?.name) {
                    toolCalls.push({ name: fc.name, args: fc.args ? { ...fc.args } : {} });
                }
            }
            if (toolCalls.length === 0)
                return null;
            await this.ensureSession().startPlan(toolCalls, this.goal);
            this.planMinted = true;
            console.info(`[armoriq] plan minted user=${this.userEmail} tools=${toolCalls.length}`);
        }
        catch (exc) {
            console.warn(`[armoriq] afterModelCallback failed: ${exc.message}`);
        }
        return null;
    }
    async beforeTool(tool, args, _toolContext) {
        const toolName = tool?.name ?? String(tool);
        try {
            const decision = await this.ensureSession().check(toolName, args ?? {}, this.userEmail);
            if (!decision.allowed) {
                const policy = decision.matchedPolicy ? ` (policy: ${decision.matchedPolicy})` : '';
                if (decision.action === 'hold') {
                    console.info(`[armoriq] HELD ${toolName} user=${this.userEmail} reason=${decision.reason} — waiting for approval...`);
                    // 30-min default, 3s → 15s exponential, matches PY hold-retry.
                    const timeoutMs = 30 * 60 * 1000;
                    const deadline = Date.now() + timeoutMs;
                    let pollIntervalMs = 3000;
                    let attempt = 0;
                    let approved = false;
                    let finalDecision = decision;
                    while (Date.now() < deadline) {
                        await new Promise((r) => setTimeout(r, pollIntervalMs));
                        pollIntervalMs = Math.min(pollIntervalMs * 1.5, 15000);
                        attempt += 1;
                        const retry = await this.ensureSession().check(toolName, args ?? {}, this.userEmail);
                        finalDecision = retry;
                        if (retry.allowed) {
                            console.info(`[armoriq] APPROVED ${toolName} attempt ${attempt}`);
                            approved = true;
                            break;
                        }
                        if (retry.action !== 'hold') {
                            console.info(`[armoriq] REJECTED ${toolName} action=${retry.action}`);
                            break;
                        }
                    }
                    if (approved)
                        return null;
                    const finalPolicy = finalDecision.matchedPolicy
                        ? ` (policy: ${finalDecision.matchedPolicy})`
                        : '';
                    this.blockedTools.add(toolName);
                    this.blockedActions.set(toolName, finalDecision.action);
                    if (finalDecision.action === 'hold') {
                        return {
                            error: `Approval timed out${finalPolicy}. Reason: ${finalDecision.reason ?? 'policy-hold'}.`,
                            armoriq_enforcement: {
                                blocked: true,
                                action: 'hold',
                                reason: finalDecision.reason,
                                matched_policy: finalDecision.matchedPolicy,
                                tool: toolName,
                                delegation_id: finalDecision.delegationId,
                            },
                        };
                    }
                    return {
                        error: `This action is not permitted${finalPolicy}. Reason: ${finalDecision.reason ?? 'policy-blocked'}.`,
                        armoriq_enforcement: {
                            blocked: true,
                            action: finalDecision.action,
                            reason: finalDecision.reason,
                            matched_policy: finalDecision.matchedPolicy,
                            tool: toolName,
                            delegation_id: finalDecision.delegationId,
                        },
                    };
                }
                // Hard block
                this.blockedTools.add(toolName);
                this.blockedActions.set(toolName, decision.action);
                console.info(`[armoriq] BLOCKED ${toolName} user=${this.userEmail} action=${decision.action} reason=${decision.reason}`);
                return {
                    error: `This action is not permitted by your organization's policy${policy}. Reason: ${decision.reason ?? 'policy-blocked'}.`,
                    armoriq_enforcement: {
                        blocked: true,
                        action: decision.action,
                        reason: decision.reason,
                        matched_policy: decision.matchedPolicy,
                        tool: toolName,
                        delegation_id: decision.delegationId,
                    },
                };
            }
        }
        catch (exc) {
            console.warn(`[armoriq] beforeToolCallback failed: ${exc.message}`);
        }
        return null;
    }
    async afterTool(tool, args, _toolContext, toolResponse) {
        const toolName = tool?.name ?? String(tool);
        try {
            if (this.blockedTools.has(toolName)) {
                const action = this.blockedActions.get(toolName) ?? 'block';
                this.blockedActions.delete(toolName);
                this.blockedTools.delete(toolName);
                if (action !== 'hold') {
                    await this.ensureSession().report(toolName, args ?? {}, toolResponse, {
                        status: 'failed',
                        errorMessage: 'Blocked by policy',
                    });
                }
                return null;
            }
            await this.ensureSession().report(toolName, args ?? {}, toolResponse);
        }
        catch (exc) {
            console.warn(`[armoriq] afterToolCallback failed: ${exc.message}`);
        }
        return null;
    }
    /**
     * Attach the three callbacks to the ADK-style agent. Save originals
     * for uninstall().
     */
    install(agent) {
        this.agent = agent;
        this.saved = {
            afterModelCallback: agent.afterModelCallback,
            beforeToolCallback: agent.beforeToolCallback,
            afterToolCallback: agent.afterToolCallback,
        };
        agent.afterModelCallback = (...args) => this.afterModel(args[0], args[1]);
        agent.beforeToolCallback = (...args) => {
            const a = args[0];
            // Require `tool` specifically — anything else means we're in legacy
            // positional mode and a is the BaseTool itself.
            if (a && typeof a === 'object' && 'tool' in a) {
                return this.beforeTool(a.tool, a.args, a.context ?? a.toolContext);
            }
            return this.beforeTool(a, args[1], args[2]);
        };
        agent.afterToolCallback = (...args) => {
            const a = args[0];
            // Require `tool` specifically — anything else means we're in legacy
            // positional mode and a is the BaseTool itself.
            if (a && typeof a === 'object' && 'tool' in a) {
                return this.afterTool(a.tool, a.args, a.context ?? a.toolContext, a.response);
            }
            return this.afterTool(a, args[1], args[2], args[3]);
        };
        return this;
    }
    uninstall(agent) {
        const a = agent ?? this.agent;
        if (!a)
            return;
        a.afterModelCallback = this.saved.afterModelCallback;
        a.beforeToolCallback = this.saved.beforeToolCallback;
        a.afterToolCallback = this.saved.afterToolCallback;
    }
}
exports.ArmorIQADKBundle = ArmorIQADKBundle;
