// @armoriq/sdk/dist/cli/state.js
"use strict";
/**
 * Local CLI state ($HOME/.armoriq/{state,cli.log}.{json,jsonl}).
 * Mirrors the layout used by armoriq_sdk/cli.py so PY and TS CLIs share the same files.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.LOG_FILE = exports.STATE_FILE = exports.ARMORIQ_DIR = void 0;
exports.loadState = loadState;
exports.saveState = saveState;
exports.clearState = clearState;
exports.appendLog = appendLog;
const fs = __importStar(require("fs"));
const os = __importStar(require("os"));
const path = __importStar(require("path"));
exports.ARMORIQ_DIR = path.join(os.homedir(), '.armoriq');
exports.STATE_FILE = path.join(exports.ARMORIQ_DIR, 'state.json');
exports.LOG_FILE = path.join(exports.ARMORIQ_DIR, 'cli.log');
function loadState() {
    try {
        if (!fs.existsSync(exports.STATE_FILE))
            return {};
        return JSON.parse(fs.readFileSync(exports.STATE_FILE, 'utf-8'));
    }
    catch {
        return {};
    }
}
function saveState(state) {
    fs.mkdirSync(exports.ARMORIQ_DIR, { recursive: true });
    fs.writeFileSync(exports.STATE_FILE, JSON.stringify(state, null, 2) + '\n', { mode: 0o600 });
}
function clearState() {
    try {
        if (fs.existsSync(exports.STATE_FILE)) {
            fs.unlinkSync(exports.STATE_FILE);
            return true;
        }
        return false;
    }
    catch {
        return false;
    }
}
function appendLog(event, details = {}) {
    try {
        fs.mkdirSync(exports.ARMORIQ_DIR, { recursive: true });
        const entry = JSON.stringify({
            timestamp: new Date().toISOString(),
            event,
            details,
        });
        fs.appendFileSync(exports.LOG_FILE, entry + '\n');
    }
    catch {
        // best-effort; never throw on log write
    }
}
