// @armoriq/sdk/dist/cli/util.js
"use strict";
/**
 * Shared CLI utilities: console formatting, prompts, env-driven endpoint
 * resolution, and the "require credentials" guard.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CLIError = exports.WARN = exports.CROSS = exports.CHECK = void 0;
exports.out = out;
exports.maskSecret = maskSecret;
exports.prompt = prompt;
exports.promptYesNo = promptYesNo;
exports.backendBase = backendBase;
exports.proxyBase = proxyBase;
exports.requireCredentials = requireCredentials;
const readline = __importStar(require("readline"));
const credentials_1 = require("../credentials");
const _build_env_1 = require("../_build_env");
exports.CHECK = '\x1b[32m✔\x1b[0m';
exports.CROSS = '\x1b[31m✘\x1b[0m';
exports.WARN = '\x1b[33m!\x1b[0m';
class CLIError extends Error {
    constructor(message) {
        super(message);
        this.name = 'CLIError';
    }
}
exports.CLIError = CLIError;
function out(message = '') {
    process.stdout.write(message + '\n');
}
function maskSecret(secret) {
    if (!secret)
        return '';
    if (secret.length <= 16)
        return secret;
    return secret.slice(0, 16) + '...';
}
async function prompt(question, defaultValue) {
    const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
    const suffix = defaultValue ? ` [${defaultValue}]` : '';
    return new Promise((resolve) => {
        rl.question(`${question}${suffix}: `, (answer) => {
            rl.close();
            resolve(answer.trim() || defaultValue || '');
        });
    });
}
async function promptYesNo(question, defaultValue = false) {
    const hint = defaultValue ? 'Y/n' : 'y/N';
    const answer = (await prompt(`${question} [${hint}]`)).toLowerCase();
    if (!answer)
        return defaultValue;
    return answer === 'y' || answer === 'yes';
}
function backendBase() {
    const explicit = process.env.BACKEND_ENDPOINT || process.env.ARMORIQ_BACKEND_URL;
    if (explicit)
        return explicit.replace(/\/+$/, '');
    return (0, _build_env_1.resolveEndpoint)('backend').replace(/\/+$/, '');
}
function proxyBase() {
    const explicit = process.env.PROXY_ENDPOINT;
    if (explicit)
        return explicit.replace(/\/+$/, '');
    return (0, _build_env_1.resolveEndpoint)('proxy').replace(/\/+$/, '');
}
function requireCredentials() {
    const creds = (0, credentials_1.loadCredentials)();
    if (!creds) {
        throw new CLIError(`Not logged in (${(0, credentials_1.getCredentialsPath)()} missing). Run \`armoriq login\` first.`);
    }
    return creds;
}
