// @armoriq/sdk/dist/cli/commands/auth.js
"use strict";
/**
 * armoriq login / logout / whoami — OAuth 2.0 device-code flow (RFC 8628).
 * Mirrors armoriq_sdk/cli_auth.py.
 *
 * The browser approval page either redirects straight back to our
 * local callback with the key, or — if the callback can't be reached —
 * we fall back to polling /auth/device/token.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.cmdLogin = cmdLogin;
exports.cmdLogout = cmdLogout;
exports.cmdWhoami = cmdWhoami;
const http = __importStar(require("http"));
const net = __importStar(require("net"));
const url_1 = require("url");
const child_process_1 = require("child_process");
const axios_1 = __importDefault(require("axios"));
const credentials_1 = require("../../credentials");
const util_1 = require("../util");
const PRODUCT_LABELS = {
    armorclaude: 'ArmorClaude',
    armorcodex: 'ArmorCodex',
    armorcopilot: 'ArmorCopilot',
    armorclaw: 'ArmorClaw',
    platform: 'ArmorIQ Platform',
    sdk: 'ArmorIQ SDK',
};
function escapeHtml(s) {
    return s.replace(/[&<>"']/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;', "'": '&#39;' })[c] ?? c);
}
function renderSuccessHtml(opts) {
    const productKey = (opts.product ?? '').toLowerCase().trim();
    const productLabel = PRODUCT_LABELS[productKey] ?? null;
    const emailHtml = opts.email
        ? `<div class="email">${escapeHtml(opts.email)}</div>`
        : '';
    const productHtml = productLabel
        ? `<div class="chip">Connected to ${escapeHtml(productLabel)}</div>`
        : '';
    const year = new Date().getFullYear();
    return `<!DOCTYPE html>
<html lang="en"><head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Authorized · ArmorIQ</title>
<link rel="icon" type="image/svg+xml" href="data:image/svg+xml;utf8,${encodeURIComponent(`<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 27 27"><path d="M6.25 16.18 9.14 11.97 11.83 15.88 14.53 11.97 17.17 15.88 20.06 11.71" stroke="#D97D55" stroke-width="2" fill="none" stroke-linecap="round"/><circle cx="13.15" cy="13.85" r="12.15" stroke="#D97D55" stroke-width="2" fill="none"/></svg>`)}">
<style>
  :root { color-scheme: light; }
  * { box-sizing: border-box; }
  html, body { height: 100%; margin: 0; }
  body {
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Inter, system-ui, sans-serif;
    background: #fafafa;
    color: #111;
    display: flex; flex-direction: column;
    min-height: 100vh;
  }
  main {
    flex: 1;
    display: flex; flex-direction: column;
    align-items: center; justify-content: center;
    gap: 28px;
    padding: 48px 32px;
  }
  .brand {
    display: flex; flex-direction: column; align-items: center; gap: 14px;
  }
  .brand svg.mark { width: 80px; height: 80px; }
  .brand .wordmark { font-weight: 700; font-size: 34px; letter-spacing: -0.025em; color: #111; }
  .card {
    background: #fff; border: 1px solid #e5e7eb; border-radius: 14px;
    padding: 40px 36px; max-width: 460px; width: 100%; text-align: center;
    box-shadow: 0 1px 2px rgba(0,0,0,0.04);
  }
  .check {
    width: 64px; height: 64px; border-radius: 999px;
    background: #dcfce7; color: #16a34a;
    display: flex; align-items: center; justify-content: center;
    margin: 0 auto 18px;
  }
  .check svg { width: 32px; height: 32px; }
  h1 { margin: 0 0 8px; font-size: 22px; letter-spacing: -0.01em; }
  p.lead { margin: 0 0 4px; color: #4b5563; font-size: 14px; line-height: 1.5; }
  .email { font-size: 13px; color: #111; font-weight: 600; margin-top: 14px; }
  .chip {
    display: inline-block; margin-top: 10px;
    padding: 4px 10px; border-radius: 999px;
    background: #f3f4f6; color: #374151;
    font-size: 12px; font-weight: 500;
  }
  .hint {
    margin-top: 20px; padding-top: 18px; border-top: 1px solid #f1f5f9;
    color: #6b7280; font-size: 12px;
  }
  footer {
    padding: 22px 32px; color: #6b7280; font-size: 12px;
    display: flex; justify-content: space-between; flex-wrap: wrap; gap: 12px;
    border-top: 1px solid #eef0f3;
  }
  footer nav a { color: #6b7280; text-decoration: none; margin-left: 16px; }
  footer nav a:hover { color: #111; text-decoration: underline; }
</style>
</head>
<body>
<main>
  <div class="brand">
    <svg class="mark" viewBox="0 0 27 27" fill="none" xmlns="http://www.w3.org/2000/svg" aria-hidden="true">
      <path d="M6.25 16.18 9.14 11.97 11.83 15.88 14.53 11.97 17.17 15.88 20.06 11.71" stroke="#D97D55" stroke-width="2" stroke-linecap="round"/>
      <circle cx="13.15" cy="13.85" r="12.15" stroke="#D97D55" stroke-width="2"/>
    </svg>
    <span class="wordmark">ArmorIQ</span>
  </div>
  <div class="card">
    <div class="check">
      <svg viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2.5" stroke-linecap="round" stroke-linejoin="round" aria-hidden="true">
        <path d="M5 13l4 4L19 7"/>
      </svg>
    </div>
    <h1>You're all set</h1>
    <p class="lead">Your device is now connected to ArmorIQ.</p>
    ${emailHtml}
    ${productHtml}
    <div class="hint">You can close this tab and return to your terminal.</div>
  </div>
</main>
<footer>
  <div>© ${year} ArmorIQ Inc.</div>
  <nav>
    <a href="https://docs.armoriq.ai" target="_blank" rel="noreferrer">Docs</a>
    <a href="https://dev.armoriq.ai" target="_blank" rel="noreferrer">Dashboard</a>
    <a href="https://armoriq.ai/privacy" target="_blank" rel="noreferrer">Privacy</a>
    <a href="https://armoriq.ai/terms" target="_blank" rel="noreferrer">Terms</a>
  </nav>
</footer>
</body></html>`;
}
function findFreePort() {
    return new Promise((resolve, reject) => {
        const server = net.createServer();
        server.unref();
        server.on('error', reject);
        server.listen(0, '127.0.0.1', () => {
            const addr = server.address();
            if (addr && typeof addr === 'object') {
                const port = addr.port;
                server.close(() => resolve(port));
            }
            else {
                reject(new Error('Could not get free port'));
            }
        });
    });
}
function startCallbackServer(port, product) {
    let resolveResult;
    const once = new Promise((resolve) => {
        resolveResult = resolve;
    });
    const server = http.createServer((req, res) => {
        const url = new url_1.URL(req.url ?? '/', `http://localhost:${port}`);
        if (url.pathname === '/callback') {
            const params = Object.fromEntries(url.searchParams.entries());
            const callbackProduct = params.product || product;
            res.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8' });
            res.end(renderSuccessHtml({
                email: params.email,
                product: callbackProduct,
                orgId: params.org_id,
            }));
            const result = {
                api_key: params.key,
                email: params.email,
                user_id: params.user_id,
                org_id: params.org_id,
            };
            if (result.api_key)
                resolveResult(result);
        }
        else {
            res.writeHead(404);
            res.end();
        }
    });
    server.listen(port, '127.0.0.1');
    return { server, once };
}
function openBrowser(url) {
    const cmd = process.platform === 'darwin'
        ? `open "${url}"`
        : process.platform === 'win32'
            ? `start "" "${url}"`
            : `xdg-open "${url}"`;
    (0, child_process_1.exec)(cmd, () => {
        /* best-effort */
    });
}
function sleep(ms) {
    return new Promise((r) => setTimeout(r, ms));
}
async function cmdLogin(args) {
    const backend = (args.backend ?? process.env.ARMORIQ_BACKEND_URL ?? (0, util_1.backendBase)()).replace(/\/+$/, '');
    const requestedOrg = (args.org ?? '').trim();
    const product = (args.product ?? process.env.ARMORIQ_PRODUCT ?? '').trim();
    // Already authenticated? Skip the browser flow unless re-auth is forced or
    // the user is switching org. Re-running `login` otherwise just re-opens the
    // browser for no reason.
    if (!args.force && !requestedOrg) {
        const existing = (0, credentials_1.loadCredentials)();
        if (existing) {
            (0, util_1.out)('');
            (0, util_1.out)(`  ${util_1.CHECK} Already logged in as \x1b[1m${existing.email || 'unknown'}\x1b[0m (org: ${existing.orgId || 'unknown'})`);
            (0, util_1.out)('  \x1b[2mRun \x1b[0m\x1b[1marmoriq login --force\x1b[0m\x1b[2m to re-authenticate, or \x1b[0m\x1b[1marmoriq logout\x1b[0m\x1b[2m first.\x1b[0m');
            (0, util_1.out)('');
            return 0;
        }
    }
    (0, util_1.out)('');
    (0, util_1.out)('  \x1b[1m\x1b[36m┃ ArmorIQ Login\x1b[0m');
    (0, util_1.out)('');
    const port = await findFreePort();
    const callbackUrl = `http://localhost:${port}/callback`;
    const { server, once: callbackOnce } = startCallbackServer(port, product);
    let dc;
    try {
        const r = await axios_1.default.post(`${backend}/auth/device/code`, { callback_url: callbackUrl }, { timeout: 10000 });
        dc = r.data;
    }
    catch (e) {
        server.close();
        (0, util_1.out)(`  ${util_1.CROSS} Failed to request device code: ${e.message}`);
        return 1;
    }
    const { device_code, user_code, verification_uri_complete } = dc;
    const interval = Number(dc.interval ?? 5) || 5;
    const expiresIn = Number(dc.expires_in ?? 600) || 600;
    const sep = verification_uri_complete.includes('?') ? '&' : '?';
    let browserUrl = `${verification_uri_complete}${sep}callback=${encodeURIComponent(callbackUrl)}`;
    if (requestedOrg)
        browserUrl += `&org=${encodeURIComponent(requestedOrg)}`;
    if (product)
        browserUrl += `&product=${encodeURIComponent(product)}`;
    (0, util_1.out)('  Opening browser...\n');
    openBrowser(browserUrl);
    (0, util_1.out)('  If the browser didn\'t open, visit:');
    (0, util_1.out)(`    \x1b[36m\x1b[1m${browserUrl}\x1b[0m\n`);
    (0, util_1.out)(`  Confirm this code in your browser: \x1b[1m${user_code}\x1b[0m\n`);
    process.stdout.write('  Waiting for authorization...');
    const deadline = Date.now() + expiresIn * 1000;
    let result;
    let lastErr;
    while (Date.now() < deadline && !result) {
        // Race callback vs poll; whichever resolves first wins.
        const callbackOrTimeout = Promise.race([
            callbackOnce.then((r) => ({ kind: 'cb', value: r })),
            sleep(interval * 1000).then(() => ({ kind: 'tick' })),
        ]);
        const tick = await callbackOrTimeout;
        if (tick.kind === 'cb' && tick.value.api_key) {
            result = tick.value;
            break;
        }
        try {
            const pr = await axios_1.default.post(`${backend}/auth/device/token`, { deviceCode: device_code }, { timeout: 10000, validateStatus: () => true });
            const data = pr.data ?? {};
            const err = data.error;
            if (err === 'authorization_pending' || err === 'slow_down')
                continue;
            if (err) {
                lastErr = data.error_description || err;
                break;
            }
            if (data.api_key) {
                result = {
                    api_key: data.api_key,
                    email: data.email,
                    user_id: data.user_id,
                    org_id: data.org_id,
                };
                break;
            }
        }
        catch {
            // network blip; keep polling
        }
    }
    server.close();
    if (!result || !result.api_key) {
        (0, util_1.out)(` ${util_1.CROSS}`);
        (0, util_1.out)(`  ${util_1.CROSS} ${lastErr ?? 'Timed out waiting for authorization. Run `armoriq login` again.'}`);
        return 1;
    }
    (0, credentials_1.saveCredentials)({
        apiKey: result.api_key,
        email: result.email ?? '',
        userId: result.user_id ?? '',
        orgId: result.org_id ?? '',
        savedAt: new Date().toISOString(),
    });
    (0, util_1.out)(` ${util_1.CHECK}`);
    (0, util_1.out)('');
    (0, util_1.out)(`  ${util_1.CHECK} Logged in as \x1b[1m${result.email ?? 'unknown'}\x1b[0m (org: ${result.org_id ?? 'unknown'})`);
    (0, util_1.out)(`  ${util_1.CHECK} API key saved to ${(0, credentials_1.getCredentialsPath)()}`);
    (0, util_1.out)('');
    return 0;
}
function cmdLogout() {
    if ((0, credentials_1.clearCredentials)()) {
        (0, util_1.out)(`  ${util_1.CHECK} Credentials removed from ${(0, credentials_1.getCredentialsPath)()}`);
    }
    else {
        (0, util_1.out)('  \x1b[2mNo credentials found — already logged out.\x1b[0m');
    }
    return 0;
}
function cmdWhoami() {
    const creds = (0, credentials_1.loadCredentials)();
    if (!creds) {
        (0, util_1.out)('  \x1b[2mNot logged in. Run `armoriq login` to authenticate.\x1b[0m');
        return 0;
    }
    (0, util_1.out)('');
    (0, util_1.out)('  \x1b[1m\x1b[36m┃ ArmorIQ Credentials\x1b[0m');
    (0, util_1.out)('');
    const keyPreview = creds.apiKey.length > 16 ? creds.apiKey.slice(0, 16) + '...' : creds.apiKey;
    (0, util_1.out)(`  Email:    \x1b[1m${creds.email || 'unknown'}\x1b[0m`);
    (0, util_1.out)(`  API Key:  \x1b[2m${keyPreview}\x1b[0m`);
    (0, util_1.out)(`  User ID:  \x1b[2m${creds.userId || 'n/a'}\x1b[0m`);
    (0, util_1.out)(`  Org ID:   \x1b[2m${creds.orgId || 'n/a'}\x1b[0m`);
    (0, util_1.out)(`  Saved at: \x1b[2m${creds.savedAt || 'n/a'}\x1b[0m`);
    (0, util_1.out)(`  File:     \x1b[2m${(0, credentials_1.getCredentialsPath)()}\x1b[0m`);
    (0, util_1.out)('');
    return 0;
}
