// @armoriq/sdk/dist/cli/commands/keys.js
"use strict";
/**
 * armoriq keys list / revoke / prune — API key management.
 * Mirrors PR #24 (TS) by symmetry; backend routes already exist:
 *   GET /api-keys                  (JwtAuthGuard — uses credentials JWT)
 *   POST /api-keys/:id/revoke      (JwtAuthGuard)
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.cmdKeysList = cmdKeysList;
exports.cmdKeysRevoke = cmdKeysRevoke;
exports.cmdKeysPrune = cmdKeysPrune;
const axios_1 = __importDefault(require("axios"));
const util_1 = require("../util");
const state_1 = require("../state");
const KEY_COUNT_WARN_THRESHOLD = 8;
async function listKeys(apiKey) {
    const url = `${(0, util_1.backendBase)()}/api-keys`;
    const response = await axios_1.default.get(url, {
        headers: { 'X-API-Key': apiKey },
        timeout: 12000,
        validateStatus: () => true,
    });
    if (response.status === 401)
        throw new util_1.CLIError('API key rejected (401). Try `armoriq login` again.');
    if (response.status >= 400)
        throw new util_1.CLIError(`Failed to list keys (HTTP ${response.status}).`);
    const data = response.data;
    if (Array.isArray(data))
        return data;
    if (Array.isArray(data?.data))
        return data.data;
    return [];
}
async function revokeKey(apiKey, keyId) {
    const url = `${(0, util_1.backendBase)()}/api-keys/${keyId}/revoke`;
    const response = await axios_1.default.post(url, {}, {
        headers: { 'X-API-Key': apiKey, 'Content-Type': 'application/json' },
        timeout: 12000,
        validateStatus: () => true,
    });
    if (response.status === 401)
        throw new util_1.CLIError('API key rejected (401). Try `armoriq login` again.');
    if (response.status >= 400) {
        throw new util_1.CLIError(`Failed to revoke key ${keyId} (HTTP ${response.status}): ${response.data?.message ?? ''}`);
    }
}
async function cmdKeysList() {
    const creds = (0, util_1.requireCredentials)();
    const keys = await listKeys(creds.apiKey);
    if (keys.length === 0) {
        (0, util_1.out)('No API keys found for this account.');
        return 0;
    }
    const nameW = Math.max('NAME'.length, ...keys.map((k) => (k.name ?? '').length));
    const idW = Math.max('ID'.length, ...keys.map((k) => k.id.length));
    (0, util_1.out)(`  ${'NAME'.padEnd(nameW)}  ${'ID'.padEnd(idW)}  STATUS    LAST USED`);
    (0, util_1.out)('  ' + '-'.repeat(nameW + idW + 30));
    for (const k of keys) {
        const name = (k.name ?? '').padEnd(nameW);
        const id = k.id.padEnd(idW);
        const status = (k.status ?? 'unknown').padEnd(8);
        const last = k.lastUsedAt ?? '-';
        (0, util_1.out)(`  ${name}  ${id}  ${status}  ${last}`);
    }
    (0, util_1.out)('');
    if (keys.length > KEY_COUNT_WARN_THRESHOLD) {
        (0, util_1.out)(`\x1b[33m!\x1b[0m You have ${keys.length} API keys. Consider \`armoriq keys prune\` to revoke unused keys.`);
    }
    (0, state_1.appendLog)('keys-list', { count: keys.length });
    return 0;
}
async function cmdKeysRevoke(args) {
    if (!args.id)
        throw new util_1.CLIError('Key id is required: `armoriq keys revoke <id>`.');
    const creds = (0, util_1.requireCredentials)();
    await revokeKey(creds.apiKey, args.id);
    (0, util_1.out)(`${util_1.CHECK} Revoked key ${args.id}.`);
    (0, state_1.appendLog)('keys-revoke', { id: args.id });
    return 0;
}
async function cmdKeysPrune(args) {
    const creds = (0, util_1.requireCredentials)();
    const keys = await listKeys(creds.apiKey);
    const now = Date.now();
    const candidates = keys.filter((k) => {
        if (k.status && k.status.toLowerCase() === 'revoked')
            return false;
        if (k.id === creds.apiKey)
            return false;
        if (k.expiresAt && new Date(k.expiresAt).getTime() < now)
            return true;
        if (k.lastUsedAt) {
            const ageDays = (now - new Date(k.lastUsedAt).getTime()) / 86400000;
            if (ageDays > 90)
                return true;
        }
        if (!k.lastUsedAt && k.createdAt) {
            const ageDays = (now - new Date(k.createdAt).getTime()) / 86400000;
            if (ageDays > 30)
                return true;
        }
        return false;
    });
    if (candidates.length === 0) {
        (0, util_1.out)('Nothing to prune — all keys are either active, recent, or already revoked.');
        return 0;
    }
    (0, util_1.out)(`Found ${candidates.length} prune candidate(s):`);
    for (const k of candidates) {
        (0, util_1.out)(`  ${k.id} ${k.name ?? ''} (last used ${k.lastUsedAt ?? 'never'})`);
    }
    if (!args.yes) {
        (0, util_1.out)('');
        (0, util_1.out)('Re-run with --yes to actually revoke these.');
        return 0;
    }
    for (const k of candidates) {
        try {
            await revokeKey(creds.apiKey, k.id);
            (0, util_1.out)(`${util_1.CHECK} Revoked ${k.id}`);
        }
        catch (e) {
            (0, util_1.out)(`\x1b[31m✘\x1b[0m Failed to revoke ${k.id}: ${e.message}`);
        }
    }
    (0, state_1.appendLog)('keys-prune', { revoked: candidates.length });
    return 0;
}
