// @armoriq/sdk/dist/cli/commands/orgs.js
"use strict";
/**
 * armoriq orgs / switch-org — list and switch ArmorIQ organizations.
 * Mirrors armoriq_sdk/cli.py:cmd_orgs / cmd_switch_org.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.cmdOrgs = cmdOrgs;
exports.cmdSwitchOrg = cmdSwitchOrg;
const axios_1 = __importDefault(require("axios"));
const credentials_1 = require("../../credentials");
const util_1 = require("../util");
const state_1 = require("../state");
const fs = __importStar(require("fs"));
async function cmdOrgs() {
    const creds = (0, util_1.requireCredentials)();
    const url = `${(0, util_1.backendBase)()}/iap/sdk/orgs`;
    let response;
    try {
        response = await axios_1.default.get(url, {
            headers: { 'X-API-Key': creds.apiKey },
            timeout: 12000,
            validateStatus: () => true,
        });
    }
    catch (e) {
        throw new util_1.CLIError(`Failed to reach ${url}: ${e.message}`);
    }
    if (response.status === 401) {
        throw new util_1.CLIError('API key rejected (401). Try `armoriq login` again.');
    }
    if (response.status >= 400) {
        throw new util_1.CLIError(`Failed to list orgs (HTTP ${response.status}).`);
    }
    const orgs = response.data?.data ?? [];
    if (orgs.length === 0) {
        (0, util_1.out)("You don't belong to any organizations.");
        return 0;
    }
    const nameW = Math.max('NAME'.length, ...orgs.map((o) => (o.name ?? '').length));
    const roleW = Math.max('ROLE'.length, ...orgs.map((o) => (o.userRole ?? '').length));
    const header = `  ${'NAME'.padEnd(nameW)}  ${'ORG_ID'.padEnd(36)}  ${'ROLE'.padEnd(roleW)}  MEMBERS`;
    (0, util_1.out)(header);
    (0, util_1.out)('  ' + '-'.repeat(header.length - 2));
    for (const org of orgs) {
        const marker = org.active ? `${util_1.CHECK} ` : '  ';
        const name = (org.name ?? '').padEnd(nameW);
        const orgId = (org.orgId ?? '').padEnd(36);
        const role = (org.userRole ?? '').padEnd(roleW);
        const members = String(org.memberCount ?? 0);
        (0, util_1.out)(`${marker}${name}  ${orgId}  ${role}  ${members}`);
    }
    (0, util_1.out)('');
    (0, util_1.out)(`Active org is marked with ${util_1.CHECK}. Switch with \`armoriq switch-org <name-or-id>\`.`);
    (0, state_1.appendLog)('orgs', { count: orgs.length });
    return 0;
}
async function cmdSwitchOrg(args) {
    const creds = (0, util_1.requireCredentials)();
    const target = (args.org ?? '').trim();
    if (!target)
        throw new util_1.CLIError('Target org (id or name) is required.');
    const url = `${(0, util_1.backendBase)()}/iap/sdk/switch-org`;
    const body = { org: target };
    if (args.keyName)
        body.keyName = args.keyName;
    let response;
    try {
        response = await axios_1.default.post(url, body, {
            headers: { 'X-API-Key': creds.apiKey, 'Content-Type': 'application/json' },
            timeout: 12000,
            validateStatus: () => true,
        });
    }
    catch (e) {
        throw new util_1.CLIError(`Failed to reach ${url}: ${e.message}`);
    }
    if (response.status === 401)
        throw new util_1.CLIError('API key rejected (401). Try `armoriq login` again.');
    if (response.status === 403)
        throw new util_1.CLIError(`You are not a member of '${target}'.`);
    if (response.status === 404)
        throw new util_1.CLIError(`No organization named '${target}' (or matching that id).`);
    if (response.status === 400) {
        throw new util_1.CLIError(response.data?.message || 'Bad request.');
    }
    if (response.status >= 400) {
        throw new util_1.CLIError(`Switch failed (HTTP ${response.status}).`);
    }
    const payload = response.data ?? {};
    const newApiKey = payload.api_key;
    const newOrgId = payload.org_id;
    const newOrgName = payload.org_name ?? target;
    if (!newApiKey || !newOrgId) {
        throw new util_1.CLIError('Switch response missing api_key or org_id.');
    }
    (0, credentials_1.saveCredentials)({
        apiKey: newApiKey,
        email: creds.email,
        userId: creds.userId,
        orgId: newOrgId,
        savedAt: new Date().toISOString(),
    });
    // Agent registration (state.json) was org-scoped — invalidate it so the
    // next step is obvious.
    const stateExisted = fs.existsSync(state_1.STATE_FILE);
    if (stateExisted)
        (0, state_1.clearState)();
    (0, util_1.out)(`${util_1.CHECK} Switched to ${newOrgName} (org_id=${newOrgId})`);
    (0, util_1.out)(`${util_1.CHECK} New API key saved.`);
    if (stateExisted) {
        (0, util_1.out)('  Previous agent registration cleared. Re-run `armoriq register` in this org.');
    }
    (0, state_1.appendLog)('switch-org', { org_id: newOrgId, org_name: newOrgName, cleared_state: stateExisted });
    return 0;
}
